File: battleships/BtlBoard.py
from BtlShip import BtlShip
from BtlMove import BtlMove
import random
from os import urandom
# ------------------------------------
# CLASS BtlBoard                     |
#                                    |
# Maintain the state of a single     |
# board                              |
# ------------------------------------
class BtlBoard( object ):

  # Create a board, size 6 or 7 - 3 ships
  #                 size 8 to 9 - 4 ships
  #                 size 10     - 5 ships
  def __init__( self, size ):
    # Initialise the random number generator
    random.seed( urandom( 128 ) )
    # Adjust size if needed
    if size < 6:
      size = 6
    elif size > 10:
      size = 10
    # Create ships
    self.__ships = [ BtlShip( 'Battleship', 4 ),
                     BtlShip( 'Destroyer', 3 ),
                     BtlShip( 'Gunship', 2 ) ]
    # Add more ships if needed 
    if size > 7:
      self.__ships.append( BtlShip( 'Submarine', 3 ) )
    if size == 10:
      self.__ships.append( BtlShip( 'Aircraft Carrier', 5 ) )
    # Create board matrix
    self.__board = [ [ 0 for x in range( size ) ] for x in range( size ) ]
    self.__size = size

  def getSize( self ):
    return self.__size

  def getShips( self ):
    return self.__ships

  # Check if all of the ships on the board have been sunk
  def isAllSunk( self ):
    for ship in self.__ships:
      if ship.isSunk is False:
        return False
    return True
  
  # Check a players move against the ships
  # on the board.
  # Returns True if a hit
  #         False if a miss
  def takeShot( self, move ):
    if isinstance( move, BtlMove ):
      for ship in self.__ships:
        if ship.isHit( move.getX(), move.getY(), True ) is True:
          return True
      return False
    else:
      raise TypeError( 'BtlBoard().takeShot() - Not a valid move object' )

  # Randomly position the availible ships on the board
  def positionShips( self ):
    overlap = True
    # Repeat until no ships overlap
    while overlap is True:
      # Pick some random positions
      for ship in self.__ships:
        onboard = False
        while onboard is False:
          x = random.randrange( 0, self.__size - 1 )
          y = random.randrange( 0, self.__size - 1 )
          d = direction = random.choice( 'v>' )
          if d == 'v':
            onboard = ( y + ship.getSize() - 1 ) < self.__size
          else: # d == '>'
            onboard = ( x + ship.getSize() - 1 ) < self.__size
        ship.setPosition( x, y, d )    
      # Check that they don't overlap
      # TODO Could this be altered sensibly to make sure they don't touch either
      overlap = False
      for ship in self.__ships:
        if overlap is True: break
        for boat in self.__ships:
          if overlap is True: break
          if ship is not boat:
            if ship.getDirection() == 'v':
              x = ship.getX()
              for y in range( ship.getY(), ship.getY() + ship.getSize() ):
                overlap = boat.isHit( x, y, False )
                if overlap is True: break
            else: # direction is '>'
              y = ship.getY()
              for x in range( ship.getX(), ship.getX() + ship.getSize() ):
                overlap = boat.isHit( x, y, False )
                if overlap is True: break

File: javascript/battleships/BtlBoard.js
// Undecided if I would prefer to change direction from v & > to 0 & 1
// The former is easier to understnd in the code, the later should be
// quicker to process

/******************************************************************************
 * Maintains state of the a game board                                       */
function BtlBoard( size ) {
  // Adjust size if needed
  if ( size < 6 ) {
    size = 6;
  } else if ( size > 10 ) {
    size = 10;
  }
  this.size = size;
  // Create the ships
  this.ships = [ new BtlShip( 'Battleship', 4 ),
                 new BtlShip( 'Destroyer', 3 ),
                 new BtlShip( 'Gunship', 2 ) ];
  if ( this.size > 7 ) {
    this.ships.push( new BtlShip( 'Submarine', 3 ) );
  }
  if ( this.size == 10 ) {
    this.ships.push( new BtlShip( 'Aircraft Carrier', 5 ) );
  }
  // Generate the board
  this.board = [];
  for ( var i = 0 ; i < this.size ; ++i ) {
    this.board[ i ] = [];
    for ( var j = 0 ; j < this.size ; ++j ) {
      this.board[ i ][ j ] = 0;
    }
  }
}
BtlBoard.prototype.getSize = function() { return this.size; };
BtlBoard.prototype.getShips = function() { return this.ships; };
// Check if all the ships on the board have been sunk
BtlBoard.prototype.isAllSunk = function() {
  var sunk = true;
  $.each( this.ships, function() {
    if ( this.isSunk() == false ) {
      sunk = false;
    }
  });
  return sunk;
};
// Check a players move against the ships on the board
BtlBoard.prototype.takeShot = function( move ) {
  for ( var i = 0 ; i < this.ships.length ; ++i ) {
    if ( this.ships[ i ].isHit( move.getX(), move.getY(), true ) ) {
      return true;
    }
  }
  return false;
};
// Randomly position the availible ships on the board so
// so they are not overlapping each other
BtlBoard.prototype.positionShips = function() {
  var onboard, overlap = true;
  var x, y, d;
  while ( overlap ) {
    // Pick some random ship positions
    for ( var i = 0 ; i < this.ships.length ; ++i ) {
      onboard = false;
      while ( false == onboard ) {
        x = Math.floor( Math.random() * ( this.size - 1 ) );
        y = Math.floor( Math.random() * ( this.size - 1 ) );
        if ( Math.round( Math.random() ) ) {
          d = 'v';
          onboard = ( y + this.ships[ i ].getSize() - 1 ) < this.size;
        } else {
          d = '>';
          onboard = ( x + this.ships[ i ].getSize() - 1 ) < this.size;
        }
      }
      this.ships[ i ].setPosition( x, y, d );
    }
    overlap = false;
    for ( var i = 0 ; i < this.ships.length ; ++i ) {
      if ( overlap ) {
        break;
      }
      for ( var j = 0 ; j < this.ships.length ; ++j ) {
        if ( overlap ) {
          break;
        }
        if ( this.ships[ i ] != this.ships[ j ] ) {
          if ( this.ships[ i ].getDirection() == 'v' ) {
            x = this.ships[ i ].getX();
            for ( y = this.ships[ i ].getY() ;
                  y < this.ships[ i ].getY() + this.ships[ i ].getSize() ;
                  ++y ) {
              if ( ( overlap = this.ships[ j ].isHit( x, y, false ) ) ) {
                break;
              }
            }
          } else { // getDirection() = '>'
            y = this.ships[ i ].getY();
            for ( x = this.ships[ i ].getX() ;
                  x < this.ships[ i ].getX() + this.ships[ i ].getSize() ;
                  ++x ) {
              if ( ( overlap = this.ships[ j ].isHit( x, y, false ) ) ) {
                break;
              }
            }
          }
        }
      }
    }
  }
};
